sw/host/opentitanlib/src/console/spi.rs

```diff
@@ -0,0 +1,51 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use std::cell::Cell;
+use std::time::Duration;
+
+use crate::io::console::ConsoleDevice;
+use crate::io::spi::{Target, Transfer};
+
+pub struct SpiConsoleDevice<'a> {
+    pub spi: &'a dyn Target,
+    pub console_next_frame_number: Cell<u32>,
+}
+
+impl<'a> SpiConsoleDevice<'a> {
+    const SPI_FRAME_HEADER_SIZE: usize = 8;
+    const SPI_MAX_DATA_LENGTH: usize = 2032;
+}
+
+impl<'a> ConsoleDevice for SpiConsoleDevice<'a> {
+    fn console_read(&self, buf: &mut [u8], _timeout: Duration) -> Result<usize> {
+        // Read the SPI console frame header.
+        let mut header = vec![0u8; SpiConsoleDevice::SPI_FRAME_HEADER_SIZE];
+        self.spi
+            .run_transaction(&mut [Transfer::Write(&[0xff; 4]), Transfer::Read(&mut header)])?;
+        let frame_number: u32 = u32::from_le_bytes(*&header[0..4].try_into().unwrap());
+        let data_len_bytes: usize = u32::from_le_bytes(*&header[4..8].try_into().unwrap()) as usize;
+        if frame_number != self.console_next_frame_number.get()
+            || data_len_bytes > SpiConsoleDevice::SPI_MAX_DATA_LENGTH
+        {
+            // This frame is junk, so we do not read the data.
+            return Ok(0);
+        }
+        self.console_next_frame_number.set(frame_number + 1);
+
+        // Read the SPI console frame data.
+        let data_len_bytes_w_pad = data_len_bytes + (4 - (data_len_bytes % 4));
+        let mut data = vec![0u8; data_len_bytes_w_pad];
+        self.spi
+            .run_transaction(&mut [Transfer::Write(&[0xff; 4]), Transfer::Read(&mut data)])?;
+
+        // Copy data to the output buffer.
+        for i in 0..(data_len_bytes as usize) {
+            buf[i] = data[i];
+        }
```

Comment:
What if `data_len_bytes` is larger than `buf.len()`?  You might consider adding a `VecDequeue` to `SpiConsoleDevice` and restructure things a bit:
```
impl SpiConsoleDevice {
  pub fn read_from_spi(...) -> Result<()> {
    // implement framing protocol
    // foreach byte received { self.queue.push_back(byte); }
  }
}

impl ConsoleDevice {
  pub fn read(self, buf: &[u8]) -> Result<usize> {
    if self.queue.len() == 0 {
      self.read_from_spi()?;
    }
    while self.queue.len() > 0 && i < buf.len() {
      buf[i] = self.queue.pop_front();
      i += 1;
    }
  }
}
```

Reply:
ah good catch, fixed.